text/0094-remove-kv-wal.md

```diff
@@ -0,0 +1,107 @@
+# Remove KV WAL
+
+- RFC PR: https://github.com/tikv/rfcs/pull/94
+- Tracking Issue: https://github.com/tikv/repo/issues/0000
+
+## Summary
+
+Disable KV WAL completely and move all states from KV DB to raft engine.
+
+## Motivation
+
+We already have all the raft logs required for recovery, it's a waste to write everything to WAL again. Although we are using cache write almost the whole time, but if a page is marked to flush, it can still trigger random latency. And if we want to use multiple rocksdb instances in one node, writing WAL can introduce a lot of random writes.
+
+## Detailed design
+
+This RFC assumes you have know how RocksDB works and recovers from restarting.
+
+### Relation between sequence number and (region ID, apply index)
+
+In RocksDB, WAL plays the role of recording both data and order. MANIFEST plays the role of recording persistent data. Now that WAL is removed, we need a new component to replace its role. Raft logs already have all the necessary data and partial order within the same peer, and the order across peers is traced by region version. So all we need to do is record the exact index KV DB belongs to.
+
+One choice is enabling atomic flush and using raft CF to record the index, but it can bring too many small compactions. It can get worse when there are many rocksdb instances and flush is invoked frequently.
+
+Another choice is storing the apply index in properties. However, on one hand, it’s not guaranteed that every time a flush will generate at least a table file. On the other hand, if all data are deleted, all files can be removed. So properties are unreliable (and verbose).
+
+The RFC proposes to build a relation between sequence number (SN) and (region ID, apply index), so apply index can be queried by querying each CF’s latest persisted SN.
+
+To build relations between SN and apply index, we need to know what SN is assigned to each write. RocksDB doesn’t have a public API to provide the information. Fortunately, exposing it is just a work of several lines.
+
+To persist the relations, we need to send them back to the raftstore thread, along with `ApplyRes`. Noticing SN is strictly monotonically increased just like log index, we can store the relations just like raft entries.
```

Comment:
How is the relation persisted, I mean, the detailed format

Reply:
I think it's an implementation details. It can be changed to best suit different engine types.